core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -117,73 +151,104 @@ public int size() {
+            continue;
+         }
+         cmd.addTombstone(tombstone);
+         if (++currentBatch == batchSize) {
+            // send and create a new command
+            rpcManager.sendTo(origin, cmd, DeliverOrder.NONE);
+            cmd = commandsFactory.buildIracTombstoneStateResponseCommand(batchSize);
+            currentBatch = 0;
+         }
+      }
+      if (currentBatch == 0) {
+         //nothing more to send
+         return;
+      }
+      rpcManager.sendTo(origin, cmd, DeliverOrder.NONE);
+   }
+
+   @Override
+   public void checkStaleTombstone(Collection<IracTombstoneInfo> tombstones) {
+      LocalizedCacheTopology topology = distributionManager.getCacheTopology();
+      for (IracTombstoneInfo tombstone : tombstones) {
+         IracTombstoneInfo data = tombstoneMap.get(tombstone.getKey());
+         if (!topology.getSegmentDistribution(tombstone.getSegment()).isPrimary() || (data != null && data.equals(tombstone))) {
+            // not a primary owner or the data is the same (i.e. it is valid)
+            continue;
+         }
+         IracTombstoneCleanupCommand cmd = commandsFactory.buildIracTombstoneCleanupCommand(tombstone);
+         rpcManager.sendToMany(topology.getSegmentDistribution(tombstone.getSegment()).writeOwners(), cmd, DeliverOrder.NONE);
+      }
+   }
+
+   // Testing purposes
    public void startCleanupTombstone() {
       iracExecutor.run();
    }
 
+   // Testing purposes
+   public void runCleanupAndWait() {
+      performCleanup().toCompletableFuture().join();
+   }
+
+   // Testing purposes
+   public boolean contains(IracTombstoneInfo tombstone) {
+      IracTombstoneInfo t = tombstoneMap.get(tombstone.getKey());
+      return t != null && t.equals(tombstone);
+   }
+
    private CompletionStage<Void> performCleanup() {
+      Function<Address, IracTombstonePrimaryCheckCommand> cmdFactory = address -> commandsFactory.buildIracTombstonePrimaryCheckCommand(batchSize);
       AggregateCompletionStage<Void> stage = CompletionStages.aggregateCompletionStage();
-      for (Map.Entry<Object, TombstoneData> entry : tombstoneMap.entrySet()) {
-         DistributionInfo info = distributionManager.getCacheTopology().getSegmentDistribution(entry.getValue().getSegment());
+      Map<Address, IracTombstonePrimaryCheckCommand> staleTombstones = new HashMap<>();
+      for (IracTombstoneInfo tombstone : tombstoneMap.values()) {
+         DistributionInfo info = distributionManager.getCacheTopology().getSegmentDistribution(tombstone.getSegment());
          if (!info.isWriteOwner()) {
             // topology changed, no longer an owner
-            remove(entry.getKey(), entry.getValue());
+            removeTombstone(tombstone);
             continue;
          }
-         if (!info.isPrimary() || iracManager.running().containsKey(entry.getKey())) {
+         if (!info.isPrimary()) {
+            if (iracManager.running().containsKey(tombstone.getKey())) {
+               // key not sent yet
```

Comment:
it won't send the update but it needs to keep track of it in case of primary owner crashes.

in this line, the node is looking for stale tombstones and, if the key exists in `IracManager`, it means the primary hasn't sent the update to the backup owners.

Reply:
Got it, could you change the comment to something like "primary hasn't sent the update to the backup owners yet" for the benefit of my future self?